### PR TITLE
Reduce scope of sync_tables

### PR DESCRIPTION
**Pull request by dougbrn**
Addresses some first steps of #245 

* Removes nobs columns from critical columns, no longer generates them from source in `_generate_object_table()`
* Sync_tables now just aligns indices, no generation of nobs information
* Adds the notion of temporary columns, which are removed when syncs occur
* Updates prune() to use an existing nobs column if provided, but otherwise calls `calc_nobs()` as part of the operation

### CODE REVIEW

File: src/tape/ensemble.py [hunk 1]
```diff
@@ -538,19 +541,38 @@ def calc_nobs(self, by_band=False, label="nobs"):
                 .reset_index()  # break up the multiindex
                 .categorize(columns=[self._band_col])  # retype the band labels as categories
                 .pivot_table(values=self._band_col, index=self._id_col, columns=self._band_col, aggfunc="sum")
-                .repartition(obj_npartitions)  # counts inherits the source partitions
             )  # the pivot_table call makes each band_count a column of the id_col row
 
+            # repartition the result to align with object
+            if self._object.known_divisions:
+                band_counts = band_counts.reset_index().set_index(self._id_col)  # ugly, but need this
```
Comment (dougbrn) [inline src/tape/ensemble.py:548]: I had to do this to resolve issues with assign not knowing divisions, and I'm really not happy about it. If anyone has ideas on how to more elegantly generate division information, let me know.

File: src/tape/ensemble.py
Comment (hombit) [inline src/tape/ensemble.py:1]: Could we also have `temporary` argument for assign? So users could add their own temporary columns.

**Review by hombit — APPROVED** [review]
Looks good, thank you!

File: src/tape/ensemble.py
Comment (dougbrn) [reply src/tape/ensemble.py:1]: Yeah I think this makes a lot of sense!

**Comment by dougbrn** [comment]
@hombit I can't seem to reply in line with the question re: why the reset_index.set_index is needed. The problem in some cases seems to be that the grouping is done on the source table, so the source partitions are preserved. And the divisions are not known unless they were known for source. The assign call seems to have issues when the divisions for object are known, but the divisions for the new pandas series to assign are not. That being said, I wonder if it'd resolve better if I tried to remove the division information from object, might give that a go

File: src/tape/ensemble.py (hunk 1 above)
Comment (dougbrn) [reply src/tape/ensemble.py:567]: The grouping is done on the source table, so the source partitions are preserved. And the divisions for the resulting dataframe are not known unless they were known for source. This leads to the problem where the assign call seems to have issues when the divisions for object are known, but the divisions for the new pandas series to assign are not. That being said, I wonder if it'd resolve better if I tried to remove the division information from object, might give that a go

File: src/tape/ensemble.py [hunk 2]
```diff
@@ -538,19 +541,38 @@ def calc_nobs(self, by_band=False, label="nobs"):
...
+                band_counts = band_counts.repartition(npartitions=self._object.npartitions)
+
             # short-hand for calculating nobs_total
             band_counts["total"] = band_counts[list(band_counts.columns)].sum(axis=1)
 
             bands = band_counts.columns.values
             self._object = self._object.assign(**{label + "_" + band: band_counts[band] for band in bands})
 
+            if temporary:
+                self._object_temp.extend([label + "_" + band for band in bands])
```
[the remark this answers is not in the document]
Comment (dougbrn) [reply src/tape/ensemble.py:579]: I think this is actually needed as a generator for a list comprehension

File: src/tape/ensemble.py (hunk 1 above)
Comment (dougbrn) [reply src/tape/ensemble.py]: The latest commit changes this to just drop divisions from object so it doesn't try to use them as part of the assign

File: src/tape/ensemble.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hombit) [reply src/tape/ensemble.py:579]: I don't think so.

Practice:
```python
In [1]: a = [-1]
In [2]: a.extend(i**2 for i in range(5))
In [3]: a
Out[3]: [-1, 0, 1, 4, 9, 16]
```

Theory:
- `list.extend()` takes any iterable, not necessary list or even a collection. That means it is not nessery to create a new list, it is enough to pass a generator
- In-line generator syntax is `(expr for item in iter)`, with round brackets. However, if in-line generator is an only argument of a function, it is allowed to omit these brackets.

I think that not having square brackets is a more pythonic way of doing the things, but of course it is a matter of taste

**Pull request merged**